macros/src/main/scala/gestalt/macros/Monadless.scala

```diff
@@ -83,120 +92,159 @@ object Transformer {
-
-        case If(Transform(monad), ifTrue, ifFalse) =>
-          val name = fresh()
-          val body = If(Ident(name), ifTrue, ifFalse)
-          Some(Nest(monad, name, body))
-
-        case If(cond, ifTrue, Some(ifFalse)) =>
-          (ifTrue, ifFalse) match {
-            case (Transform(ifTrue), Transform(ifFalse)) =>
-              Some(If(cond, ifTrue, Some(ifFalse)))
-            case (Transform(ifTrue), ifFalse) =>
-              val elsep = Apply(Resolve.apply(tree.pos), List(ifFalse))
-              Some(If(cond, ifTrue, Some(elsep)))
-            case (ifTrue, Some(Transform(ifFalse))) =>
-              val ifp = Apply(Resolve.apply(tree.pos), List(ifTrue))
-              Some(If(cond, ifp, Some(ifFalse)))
-            case (ifTrue, ifFalse) =>
-              None
-          }
+        case block @ Block(stats, expr) => TransformBlock.unapply(stats.toList :+ expr)(block.tpe)
 
+        case If(cond, ifTrue, ifFalse) => TransformIf.unapply(tree)
 
         case q"$fun[$tp]($v)" if isUnlift(fun.tpe) => Some(v)
 
-        case tree: Tree =>
-          val unlifts = collection.mutable.ListBuffer.empty[(TermTree, String, TypeTree)]
-          val newTree: TermTree =
+        case tree =>
+          val unlifts = collection.mutable.ListBuffer.empty[(tpd.Tree, Symbol, tpd.Tree)]
+          val newTree =
             transform(tree) {
-              case q"$fun[$tp]($v)" if isUnlift(fun.tpe) =>
+              case tree @ q"$fun[$tp]($v)" if isUnlift(fun.tpe) =>
                 val name = fresh()
-                val splice = wrap(tp)
-                unlifts += ((v, name, splice))
-                Ident(name)
-            }.asInstanceOf[TermTree]
+                val dummy = newValSymbol(name, tree.tpe)
+                unlifts += ((v, dummy, tp))
+                Ident(dummy)
+            }
 
           unlifts.toList match {
             case List() => None
-            case List((tree, name, _)) =>
-              val fun = Function(toParam(name) :: Nil, newTree)
-              Some(q"${Resolve.map(tree.pos, tree)}($fun)")
+            case List((tree, dummy, tpt)) =>
+              val res = rewrite(tree, dummy.name, tpt.tpe, newTree.tpe, flat = false) { refs =>
+                newTree.subst(dummy :: Nil, refs.head.symbol.get :: Nil)
+              }
+
+              Some(res)
             case unlifts =>
-              val (trees, names, types) = unlifts.unzip3
+              val (trees, dummies, types) = unlifts.unzip3
               val list = fresh("list")
-              val iterator = fresh("iterator")
-              val collect = q"${Resolve.collect(tree.pos)}(scala.List(..${trees.toSeq}))"
+              val scalaList = Ident(Type.termRef("scala.collection.immutable.List").symbol.get).select("apply")
+              val seqLiteral = SeqLiteral(trees, trees.head.tpe.widen)
+              val arg = scalaList.appliedToTypes(trees.head.tpe.widen.toTree).appliedTo(seqLiteral)
```

Comment:
Generally I feel easier and consistent working with constructors directly for typed trees.